rfc/rfc-87/rfc-87.md

```diff
@@ -24,9 +24,329 @@
+
+## Background
+
+Currently, Flink writes in Hudi rely heavily on Avro for serialization, deserialization, and in-memory storage, particularly in DataBucket and AppendHandle.
+This approach introduces unnecessary GC pressure, frequent avroToBytes/bytesToAvro conversions, and overall performance bottlenecks.
+
+With RFC-84, Avro was eliminated from Flink’s pre-write operators. However, the write path still suffers from excessive Avro-based processing, leading to inefficiencies.
+StreamWriteFunction now receives HoodieFlinkInternalRow, which already encapsulates Flink’s native RowData, making it possible to remove Avro entirely from the write phase.
+
+## Implementation
+
+The keys ideas of Flink native implementation is considered in rework of DataBucket and Flink write handles.
+This implementation doesn't touch neither bulk_insert nor append mode and works for all index types.
+It changes only StreamWriteFunction logic.
+
+### DataBucket
+
+To reduce GC pressure and eliminate Avro usage DataBucket could store RowData records in BinaryInMemorySortBuffer.
+Instead of getRecords() DataBucket could provide getIterator() method, which will be passed to Flink write handles.
+MutableObjectIterator is Flink's @Internal class, so it's better to create Hudi internal Iterator which will incapsulate MutableObjectIterator's logic.  
+
+```Java
+protected static class DataBucket {
+  private final BinaryInMemorySortBuffer sortBuffer;
+  private final BufferSizeDetector detector;
+
+  private DataBucket(Double batchSize, RowType rowType) {
+    this.sortBuffer = SortBufferFactory.createInMemorySortBuffer(rowType);
+    this.detector = new BufferSizeDetector(batchSize);
+  }
+  
+  public MutableObjectIterator<BinaryRowData> getIterator() {
+    return this.sortBuffer.getIterator();
+  }
+  
+  public boolean isEmpty() {
+    return sortBuffer.isEmpty();
+  }
+  
+  public void reset() {
+    this.sortBuffer.reset();
+    this.detector.reset();
+  }
+
+  public void add(HoodieFlinkInternalRow record) throws IOException {
+    this.sortBuffer.write(record.getRowDataWithMetadata());
+  }
+}
+```
+
+BinaryInMemorySortBuffer could ve provided by SortBufferFactory implementation.
+
+```Java
+class SortBufferFactory {
+
+  private static final int DEFAULT_PAGE_SIZE = 65536;
+
+  static BinaryInMemorySortBuffer createInMemorySortBuffer(RowType rowType) {
+    NormalizedKeyComputer normalizedKeyComputer = new NormalizedRecordKeyComputer();
+    RecordComparator keyComparator = new NormalizedRecordKeyComputer();
+
+    RowDataSerializer serializer = new RowDataSerializer(rowType);
+    BinaryRowDataSerializer binarySerializer = new BinaryRowDataSerializer(rowType.getFieldCount());
+
+    MemorySegmentPool unlimitedMemoryPool = new UnlimitedHeapMemorySegmentPool(DEFAULT_PAGE_SIZE);
+
+    return BinaryInMemorySortBuffer.createBuffer(normalizedKeyComputer, serializer, binarySerializer,
+        keyComparator, unlimitedMemoryPool);
```

Comment:
For memory pool. There can be options for users, i.e., heap based memory pool and managed memory based pool provided by flink MemoryManager, like `LazyMemorySegmentPool`.

Reply:
With flink managed memory based memory pool, we can entirely free the writer buffer from the influence of JVM GC. But it may need users to update the memory config of flink job, which is not quite compatible. So maybe we can use heap based memory pool by default, and provide the managed memory as an option, WDYT

Reply:
I think it's possible. Usage of factory provides ability to use different memory pools. For the first stage, i think it's ok to use only heap based memory pool, then add managed memory pool.

Reply:
Not sure if we can we just check the Flink runtime configuration to decide which one to use.

Reply:
To use managed memory, operators may first declare certain size of memory by `Transformation#declareManagedMemoryUseCaseAtOperatorScope`, otherwise, may lead to unexpected behavior because of memory contention, such as failure of state operations.